### PR TITLE
chore: Yarn Workspaces

### PR DESCRIPTION
**Pull request by jmelberg-okta**
### Description

Migrates the repo to use [Yarn Workspaces](https://yarnpkg.com/blog/2017/08/02/introducing-workspaces/) for managing dependencies as opposed to [Lerna](https://lernajs.io/).

This provides the following benefits:

  - Run `yarn install` once to install all workspace dependencies
  - Dependencies are linked together, allowing multiple workspaces to depend on one another

For readability, this PR is broken into the following commits:
  - [`7a1da4c`](https://github.com/okta/okta-oidc-js/pull/314/commits/7a1da4ccd5219da85feb90ce2b0a13a49a340cda) - Migrate `jwt-verifier` to use `yarn` instead of `npm`.
  - [`99f37d7`](https://github.com/okta/okta-oidc-js/pull/314/commits/99f37d7e6ad88fe02edb3cd5e3af1b29f3908812) - Migrate `oidc-middleware` to use `yarn` instead of `npm`.
  - [`a5d5dd`](https://github.com/okta/okta-oidc-js/pull/314/commits/a5d5dddef67c0f58122f5e715752a6444b1aa502) - Migrate `okta-angular` to use `yarn` instead of `npm`.
  - [`bfb2c45`](https://github.com/okta/okta-oidc-js/pull/314/commits/bfb2c4599b5ae94c2b548c3f70f11f1f5027a080) - Migrate `okta-react` to use `yarn` instead of `npm`.
  - [`93ea299`](https://github.com/okta/okta-oidc-js/pull/314/commits/93ea2990c3cbe664e6e6124faec884ae6ebe5923) - Migrate `okta-react-native` to use `yarn` instead of `npm`.
  - [`1ecc7e2`](https://github.com/okta/okta-oidc-js/pull/314/commits/1ecc7e2f81c512dd5b441ee708ea7d2b87184709) - Migrate `okta-vue` to use `yarn` instead of `npm`.
  - [`83aab80`](https://github.com/okta/okta-oidc-js/pull/314/commits/83aab806936bbeb26c8179d796a20a7a141a2135) - Use yarn workspaces

### Resolves

- This is the necessary ground work for #297

### CODE REVIEW

File: packages/okta-react/package.json [hunk 1]
```diff
@@ -4,17 +4,17 @@
   "description": "React support for Okta",
   "main": "./dist/index.js",
   "scripts": {
-    "build": "rimraf dist/ && babel src -d dist",
-    "build:harness": "npm --prefix test/e2e/harness install",
+    "build": "rimraf dist/ && yarn build:package-info && babel src -d dist",
```
Comment (manueltanzi-okta) [inline packages/okta-react/package.json]: What is the reason for this change? Should we do that also for the other frameworks?

File: packages/okta-react/package.json [hunk 2]
```diff
@@ -72,6 +72,7 @@
   "jest": {
     "setupFiles": [
       "./test/jest/setup.js"
-    ]
+    ],
+    "testURL": "http://localhost/"
```
Comment (manueltanzi-okta) [inline packages/okta-react/package.json]: I think this was fixed in Jest 23.5.0 https://github.com/facebook/jest/issues/6766#issuecomment-412516712, maybe let's update Jest instead

File: packages/jwt-verifier/README.md [hunk 3]
```diff
@@ -17,10 +17,10 @@ To learn more about verification cases and Okta's tokens please read [Working Wi
 
 ## How to use
 
-Install this library from npm:
+We use Yarn as our node package manager. To install Yarn, check out their [install documentation](https://yarnpkg.com/en/docs/install).
```
Comment (rchild-okta) [inline packages/jwt-verifier/README.md]: I'm not sure we want to expose yarn at this top-level - a consumer of the package can use yarn or npm to install (usually, you'll see documentation that just uses npm or shows commands for installs with either npm or yarn). Yarn only comes into play when you're talking about development, which could be summed up in one section in the root README (or linked from the root README).

File: packages/oidc-middleware/package.json [hunk 4]
```diff
@@ -32,7 +32,7 @@
     "express": "^4.16.3",
     "lodash": "^4.17.5",
     "negotiator": "^0.6.1",
-    "openid-client": "^2.1.0",
+    "openid-client": "2.1.0",
```
Comment (rchild-okta) [inline packages/oidc-middleware/package.json]: Interesting - is this intentional because a newer version is not compatible?

Edit: Heh, unfortunately the comments you left don't line up anymore with the shas. But, did see your answer, 👍

File: packages/okta-angular/README.md [hunk 5]
```diff
@@ -30,7 +30,11 @@ This library is tested against the latest version of Angular (currently 6), and
 This library is available through [npm](https://www.npmjs.com/package/@okta/okta-angular). To install it, simply add it to your project:
 
 ```bash
+# npm
```
Comment (rchild-okta) [inline packages/okta-angular/README.md]: This is closer to what I was talking about in the other comment (especially with the yarn part only mentioned in the development section)

File: packages/okta-angular/README.md [hunk 6]
```diff
@@ -284,23 +288,31 @@ Returns the stored URI and query parameters stored when the `OktaAuthGuard` and/
...
+   - `yarn install`
+3. Navigate into the `okta-angular` package:
+   - `cd packages/okta-angular`
 4. Make your changes to `okta-angular/src/`
+5. Set the following environment variables:
+   - `ISSUER` - your authorization server
+   - `CLIENT_ID` - the client id of your app
+   - `USERNAME` - username of org user, needed if you want to run tests
+   - `PASSWORD` - password of org user, needed if you want to run tests
+6. Start a sample server:
```
Comment (rchild-okta) [inline packages/okta-angular/README.md]: nit: start *the* sample server

File: packages/okta-angular/README.md [hunk 7]
```diff
@@ -284,23 +288,31 @@ Returns the stored URI and query parameters stored when the `OktaAuthGuard` and/
...
+   - `yarn start`
```
Comment (rchild-okta) [inline packages/okta-angular/README.md]: Personal preference, but I like using blocks for commands (and would help distinguish from the inline `ISSUER`, etc that's in the rest of these instructions), i.e.

6. Start the sample server:
    ```bash
    yarn start
    ```

File: packages/okta-angular/README.md [hunk 8]
```diff
@@ -284,23 +288,31 @@ Returns the stored URI and query parameters stored when the `OktaAuthGuard` and/
...
 
 | Command        | Description                        |
 | -------------- | ---------------------------------- |
-| `npm start`    | Start the sample app using the SDK |
-| `npm test`     | Run integration tests              |
-| `npm run lint` | Run eslint linting tests           |
-| `npm run docs` | Generate typedocs                  |
+| `yarn start`    | Start the sample app using the SDK |
+| `yarn test`     | Run integration tests              |
+| `yarn lint` | Run eslint linting tests           |
```
Comment (rchild-okta) [inline packages/okta-angular/README.md]: Nit - formatting. I use the [markdown-table-formatter](https://marketplace.visualstudio.com/items?itemName=josa.markdown-table-formatter) VSCode plugin for this.

File: packages/okta-angular/package.json [hunk 9]
```diff
@@ -9,18 +9,18 @@
...
-    "start": "npm run --prefix test/e2e/harness/ start",
+    "test": "yarn lint && yarn test:e2e && yarn test:unit",
+    "test:unit": "yarn --cwd test/e2e/harness/ test",
+    "test:e2e": "yarn --cwd test/e2e/harness/ e2e",
+    "start": "yarn --cwd test/e2e/harness/ start",
     "docs": "typedoc --options typedoc.json --exclude '{**/*.spec.ts,**/test/**}' ./src/",
-    "lint": "npm run --prefix test/e2e/harness/ lint",
-    "build:dependencies": "npm run ngc && npm pack && cd test/e2e/harness && npm install ../../../okta-okta-angular-$npm_package_version.tgz && npm install"
+    "lint": "yarn --cwd test/e2e/harness/ lint",
+    "build:dependencies": "yarn ngc && yarn pack && cd test/e2e/harness && yarn add file:../../../okta-okta-angular-v$npm_package_version.tgz && yarn install"
```
Comment (rchild-okta) [inline packages/okta-angular/package.json]: What is the `yarn add file` part for?

File: packages/okta-angular/test/e2e/harness/README.md [hunk 10]
```diff
@@ -21,6 +21,6 @@ export PASSWORD={password}
 ```bash
 # At project root
 cd packages/okta-angular
-npm install
-npm test
+yarn install
```
Comment (rchild-okta) [inline packages/okta-angular/test/e2e/harness/README.md]: Is this necessary? I.e. in general with workspaces, you run a `yarn install` in the root dir, and there's only one yarn.lock that sits at the root. Couple things to revisit over this entire PR:
- All the places where we're referencing `yarn installs` in the package documentation
- All the packages that have their own `yarn.lock`

File: packages/okta-angular/test/e2e/harness/package.json [hunk 11]
```diff
@@ -22,14 +22,14 @@
     "@angular/platform-browser": "^6.0.7",
     "@angular/platform-browser-dynamic": "^6.0.7",
     "@angular/router": "^6.0.7",
-    "@okta/okta-angular": "file:../../../okta-okta-angular-1.0.2.tgz",
+    "@okta/okta-angular": "file:../../../okta-okta-angular-v1.0.3.tgz",
```
Comment (rchild-okta) [inline packages/okta-angular/test/e2e/harness/package.json]: Should probably talk about this in a separate discussion, but am wondering if there's a better way to do this, i.e.
- There's a local test harness package
- There's a published library package that it tests
- Is the best way to set the dependency in the test harness to depend on a local reference to a hardcoded version tarball path?

File: packages/okta-react/README.md [hunk 12]
```diff
@@ -338,12 +343,12 @@ Parses tokens from the url and stores them.
...
 ## Commands
 
 | Command        | Description                        |
 | -------------- | ---------------------------------- |
-| `npm start`    | Start the sample app using the SDK |
-| `npm test`     | Run integration tests              |
-| `npm run lint` | Run eslint linting tests           |
+| `yarn start`    | Start the sample app using the SDK |
+| `yarn test`     | Run integration tests              |
+| `yarn run lint` | Run eslint linting tests           |
```
Comment (rchild-okta) [inline packages/okta-react/README.md]: nit: yarn ~~run~~ lint

File: packages/okta-react-native/__tests__/token-client-util.test.js [hunk 13]
```diff
@@ -11,10 +11,7 @@
  */
 
 import {
-  requests,
```
Comment (rchild-okta) [inline packages/okta-react-native/__tests__/token-client-util.test.js]: Probably something for later, but is there linting for this package?

File: packages/okta-vue/README.md [hunk 14]
```diff
@@ -264,21 +268,27 @@ Parses the tokens returned as hash fragments in the OAuth 2.0 Redirect URI.
 
 ## Development
 
+We use Yarn as our node package manager. To install Yarn, check out their [install documentation](https://yarnpkg.com/en/docs/install).
+
 1. Clone the repo:
     - `git clone git@github.com:okta/okta-oidc-js.git`
-2. Navigate into the `okta-vue` package:
+2. Install the dependencies:
```
Comment (rchild-okta) [inline packages/okta-vue/README.md]: Similar to the other comment, but am curious if we want to keep reiterating this in each of the packages, or just create one development setup in the root README (or linked off the root README) that talks about cloning/monorepos/install, and just refer to it in these individual packages.

File: .travis.yml [hunk 15]
```diff
@@ -10,16 +10,16 @@ addons:
...
 - sh ./scripts/runTests.sh
 - bash ./scripts/snyk.sh
 
 install:
-- npm install
+- yarn install
 
 before_install:
-- "npm install -g npm@5.6.0"
+- curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version 1.8.0
```
Comment (rchild-okta) [inline .travis.yml]: Is there a reason to use 1.8.0 instead of latest stable 1.9.4?

File: scripts/publish_external.sh [hunk 16]
```diff
@@ -1,11 +1,8 @@
 # undo permissions change on scripts/publish_external.sh
 git checkout -- scripts
 
-# Use newer, faster npm
-npm install -g npm@5.0.3
-
 # Install required dependencies
-npm install -g lerna
+yarn global add lerna
```
Comment (rchild-okta) [inline scripts/publish_external.sh]: Is there a reason to add this globally? It's already installed as a devDependency in the root package.

Also, is the following line to bootstrap no longer necessary?

File: scripts/setup.sh [hunk 17]
```diff
@@ -16,12 +16,9 @@ git config --global user.email "oktauploader@okta.com"
 git config --global user.name "oktauploader-okta"
 
 # Install required dependencies
-npm install -g lerna
+yarn global add lerna
```
Comment (rchild-okta) [inline scripts/setup.sh]: Same comment here

File: scripts/setup.sh [hunk 18]
```diff
@@ -16,12 +16,9 @@ git config --global user.email "oktauploader@okta.com"
...
 
-# We are skipping react-native until we have a solution for the problem described here: https://github.com/expo/expo/issues/1767
-# If expo will not resolve then we will have to find another workaround
-
-if ! lerna bootstrap --ignore "@okta/okta-react-native"; then
+if ! yarn install; then
   echo "lerna bootstrap failed! Exiting..."
```
Comment (rchild-okta) [inline scripts/setup.sh]: nit: different error message (no more bootstrap)

File: packages/okta-react-native/__tests__/token-client-util.test.js (hunk 13 above)
Comment (jmelberg-okta) [reply packages/okta-react-native/__tests__/token-client-util.test.js]: Unfortunately not within the `okta-react-native` package.

I'm planning to bring `estlint` out to the root of the repo, so we can run it against all packages in a future PR.

File: packages/okta-angular/test/e2e/harness/package.json (hunk 11 above)
Comment (jmelberg-okta) [reply packages/okta-angular/test/e2e/harness/package.json]: Yep. I was thinking about the same thing, where this PR lays the ground work to finding a much better way to run our tests.

With the shared workspace model, we can now have a shared test-suite that uses the same `page-objects` and specs. Right now, we're duplicating a significant amount of code.

Re: the `tgz` import:

This was a decision I made when originally publishing this library, where we would try to emulate a consumer app as close as possible. Over the last year, I've discovered that there are much better ways of doing this. 😄

File: packages/okta-react/package.json (hunk 1 above)
Comment (jmelberg-okta) [reply packages/okta-react/package.json]: Good question - This mirrors what we have with `okta-angular`, where we run `build:package-info` before we compile it.

This change keeps us in sync

File: scripts/publish_external.sh [hunk 19]
```diff
@@ -1,14 +1,8 @@
...
 
-# Use newer, faster npm
-npm install -g npm@5.0.3
-
-# Install required dependencies
-npm install -g lerna
-
-if ! lerna bootstrap; then
-  echo "lerna bootstrap failed! Exiting..."
+if ! yarn install; then
```
Comment (manueltanzi-okta) [inline scripts/publish_external.sh]: This fails on Bacon, I created a JIRA for eng services for a similar problem in okta-signin-widget https://oktainc.atlassian.net/browse/OKTA-190445 in case we cannot find a solution.

File: README.md [hunk 20]
```diff
@@ -3,13 +3,141 @@ This is a monorepo that contains Okta's OpenID Connect JavaScript resources.
...
+
+### Prerequisites
+
+#### Create a SPA
+
+1. Applications > Add Application
+2. Select SPA
+3. Add the following **login redirect URI**:
+    - `http://localhost:8080/implicit/callback`
+    - `http://localhost:3000/implicit/callback`
```
Comment (manueltanzi-okta) [inline README.md]: Is there a reason we need 2 ports for the samples/tests? Can we refactor to use only `8080`, which is the default in the Developer Console?

Comment (jmelberg-okta) [reply README.md]: Yep. This is one of the few items that are going to be picked up in a follow-up PR (related to #297)

Comment (manueltanzi-okta) [reply README.md]: Cool!
